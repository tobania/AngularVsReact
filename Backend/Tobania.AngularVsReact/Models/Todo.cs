
using System;
using System.Collections.Generic;
using System.Linq;

namespace Tobania.AngularVsReact.Models
{
    public class Todo
    {
        public Guid Id { get; set; }
        public string Title { get; set; }
        public DateTime CreatedDate { get; set; }

        public ICollection<TodoItem> TodoItems { get; set; } = new List<TodoItem>();
        public Todo() { }

        public Todo(string title)
        {
            Id = Guid.NewGuid();
            Title = title;
            CreatedDate = DateTime.UtcNow;
        }

        public void Update(string title) => Title = title;

        public TodoItem AddTodoItem(string text)
        {
            var item = new TodoItem(this, text);
            TodoItems.Add(item);

            return item;
        }

        public TodoItem GetTodoItem(Guid id) => TodoItems.FirstOrDefault(i => i.Id == id);
    }
}