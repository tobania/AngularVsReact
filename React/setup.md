# React Setup

## Prerequirements
- nodejs
- Backend runs in the background and is accessable

# Setup
Navigate to the frontend root (src) and execute the following commands:
npm install
npm start
