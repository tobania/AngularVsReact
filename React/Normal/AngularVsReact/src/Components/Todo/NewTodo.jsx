import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import CreateTodo from '../../HttpServices/Todos/CreateTodoService';

class NewTodo extends PureComponent {
  constructor() {
    super();
    this.state = { title: '' };

    this.handleChange = this.handleChange.bind(this);
    this.addTodo = this.addTodo.bind(this);
  }

  handleChange(event) {
    this.setState({ title: event.target.value });
  }

  addTodo() {
    const { title } = this.state;
    const { addCreatedTodo } = this.props;
    CreateTodo({ title }, addCreatedTodo);
  }

  render() {
    const { title } = this.state;

    return (
      <div className="col-md-3">
        <div className="card shadow mb-4">
          <div className="card-header">
            <div className="row">
              <div className="col-9">
                <input type="text" className="form-control" value={title} onChange={this.handleChange} />
              </div>
              <div className="offset-1 col-2">
                <button type="button" className="btn btn-success float-right" onClick={this.addTodo}>Save</button>
              </div>
            </div>
          </div>
          <div className="card-body" />
        </div>
      </div>
    );
  }
}

NewTodo.propTypes = {
  addCreatedTodo: PropTypes.func.isRequired,
};

NewTodo.defaultProps = {
};

export default NewTodo;
