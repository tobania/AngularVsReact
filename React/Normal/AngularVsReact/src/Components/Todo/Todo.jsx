import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import TodoItem from './Item/TodoItem';
import NewTodoItem from './Item/NewTodoItem';

class Todo extends PureComponent {
  render() {
    const { identifier, todoItems, title, addCreatedTodoItem, toggleTodoItem } = this.props;
    return (
      <div className="col-md-3">
        <div className="card shadow mb-4">
          <div className="card-header">
            <span>{title}</span>
          </div>
          <div className="card-body">
            {
              todoItems.map(todoItem => (
                <TodoItem
                  key={todoItem.identifier}
                  todoId={identifier}
                  toggleTodoItem={toggleTodoItem}
                  {...todoItem}
                />
              ))
            }
            <NewTodoItem addCreatedTodoItem={addCreatedTodoItem} identifier={identifier} />
          </div>
        </div>
      </div>
    );
  }
}

Todo.propTypes = {
  identifier: PropTypes.string.isRequired,
  title: PropTypes.string,
  todoItems: PropTypes.arrayOf(
    PropTypes.shape({
      identifier: PropTypes.string,
      text: PropTypes.string,
    }),
  ),
  addCreatedTodoItem: PropTypes.func.isRequired,
  toggleTodoItem: PropTypes.func.isRequired,
};

Todo.defaultProps = {
  title: '',
  todoItems: [],
};

export default Todo;
