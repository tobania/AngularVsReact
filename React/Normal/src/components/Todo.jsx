import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Task from './Task';

class Todo extends Component {
  render() {
    const { tasks, title } = this.props;
    return (
      <div className="col-md-3">
        <div className="card shadow mb-4">
          <div className="card-header">
            <span>{title}</span>
          </div>
          <div className="card-body">
            {
              tasks.map(task => (
                <Task task={task} />
              ))
            }
            <div className="row mt-4">
              <div className="col-10">
                <input type="text" className="form-control" />
              </div>
              <div className="col-2">
                <div className="btn btn-success float-right">+</div>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

Todo.propTypes = {
  title: PropTypes.string,
  tasks: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number,
      name: PropTypes.string,
    }),
  ),
};

Todo.defaultProps = {
  title: '',
  tasks: [],
};

export default Todo;
