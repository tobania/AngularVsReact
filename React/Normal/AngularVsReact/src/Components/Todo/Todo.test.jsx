import React from 'react';
import renderer from 'react-test-renderer';
import Todo from './Todo';

describe('Todo', () => {
  it('gets loaded correctly', () => {
    const todoItems = [{
      identifier: 'identifier1',
      text: 'text1',
    },
    {
      identifier: 'identifier2',
      text: 'text2',
    }];

    const component = renderer.create(
      <Todo identifier="identifier" title="title" todoItems={todoItems} addCreatedTodoItem={jest.fn} toggleTodoItem={jest.fn} />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
