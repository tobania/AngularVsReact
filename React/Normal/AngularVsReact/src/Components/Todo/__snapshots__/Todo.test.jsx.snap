// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo gets loaded correctly 1`] = `
<div
  className="col-md-3"
>
  <div
    className="card shadow mb-4"
  >
    <div
      className="card-header"
    >
      <span>
        title
      </span>
    </div>
    <div
      className="card-body"
    >
      <li
        className="list-group-item"
      >
        <div
          className="form-check"
        >
          <input
            checked={false}
            className="form-check-input"
            id="identifier1"
            onChange={[Function]}
            type="checkbox"
          />
          <label
            className="form-check-label"
            htmlFor="identifier1"
          >
            text1
          </label>
        </div>
      </li>
      <li
        className="list-group-item"
      >
        <div
          className="form-check"
        >
          <input
            checked={false}
            className="form-check-input"
            id="identifier2"
            onChange={[Function]}
            type="checkbox"
          />
          <label
            className="form-check-label"
            htmlFor="identifier2"
          >
            text2
          </label>
        </div>
      </li>
      <div
        className="row mt-4"
      >
        <div
          className="col-10"
        >
          <input
            className="form-control"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <div
          className="col-2"
        >
          <button
            className="btn btn-success float-right"
            onClick={[Function]}
            type="button"
          >
            +
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
